superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx

```diff
@@ -0,0 +1,481 @@
+        },
+      },
+    }));
+  }
+
+  onExpandFilterField(expandedFilterIds) {
+    this.setState(() => ({
+      expandedFilterIds,
+    }));
+  }
+
+  onSearchInputChange(e) {
+    this.setState({ searchText: e.target.value }, this.filterTree);
+  }
+
+  onChangeFilterField(activeKey) {
+    if (this.allfilterFields.includes(activeKey)) {
+      this.setState({ activeKey });
+    }
+  }
+
+  onToggleEditMode() {
+    const { activeKey, isSingleEditMode, checkedFilterFields } = this.state;
+    const { dashboardFilters } = this.props;
+    if (isSingleEditMode) {
+      // single edit => multi edit
+      this.setState(
+        {
+          isSingleEditMode: false,
+          checkedFilterFields: [activeKey],
+          filterFieldNodes: getFilterFieldNodesTree({
+            dashboardFilters,
+            isSingleEditMode: false,
+          }),
+        },
+        () => this.onCheckFilterField([activeKey]),
+      );
+    } else {
+      // multi edit => single edit
+      const nextActiveKey =
+        checkedFilterFields.length === 0
+          ? this.defaultFilterKey
+          : checkedFilterFields[0];
+
+      this.setState(() => ({
+        isSingleEditMode: true,
+        activeKey: nextActiveKey,
+        checkedFilterFields: [activeKey],
+        filterFieldNodes: getFilterFieldNodesTree({
+          dashboardFilters,
+          isSingleEditMode: true,
+        }),
+      }));
+    }
+  }
+
+  onClose() {
+    this.props.onCloseModal();
+  }
+
+  onSave() {
+    const { filterScopeMap } = this.state;
+
+    console.log(
+      'i am current state',
+      this.allfilterFields.reduce(
+        (map, key) => ({
+          ...map,
+          [key]: filterScopeMap[key].checked,
+        }),
+        {},
+      ),
+    );
+    this.props.onCloseModal();
+  }
+
+  filterTree() {
+    // Reset nodes back to unfiltered state
+    if (!this.state.searchText) {
+      this.setState(prevState => {
+        const { activeKey, filterScopeMap } = prevState;
+        const updatedEntry = {
+          ...filterScopeMap[activeKey],
+          nodesFiltered: filterScopeMap[activeKey].nodes,
+        };
+        return {
+          filterScopeMap: {
+            ...filterScopeMap,
+            [activeKey]: updatedEntry,
+          },
+        };
+      });
+
+      return;
+    }
+
+    const updater = prevState => {
+      const { activeKey, filterScopeMap } = prevState;
+      const nodesFiltered = filterScopeMap[activeKey].nodes.reduce(
+        this.filterNodes,
+        [],
+      );
+      const updatedEntry = {
+        ...filterScopeMap[activeKey],
+        nodesFiltered,
+      };
+      return {
+        filterScopeMap: {
+          ...filterScopeMap,
+          [activeKey]: updatedEntry,
+        },
+      };
+    };
+
+    this.setState(updater);
+  }
+
+  filterNodes(filtered, node) {
+    const { searchText } = this.state;
+    const children = (node.children || []).reduce(this.filterNodes, []);
```

Comment:
Theoretically all recursion can be written in loop. But here recursion is really easier than regular loop. Here we use array type to represent a **Tree** data structure, because we really care about the hierarchy of the charts and its parents. The filter function is to find leaf node that matches search keyword, as well as all its parents tab up to root component. Recursive traverse is the best way to do this job.